[JSC] Add support for Atomics.waitAsync

hyjorc1 opened the pull request:
#### 7e3fb31d598706ca0d25f05030fa41fbfb458566
<pre>
[JSC] Add support for Atomics.waitAsync
<a href="https://bugs.webkit.org/show_bug.cgi?id=241414">https://bugs.webkit.org/show_bug.cgi?id=241414</a>
rdar://94655073

Reviewed by Yusuke Suzuki.

Atomics.waitAsync() waits asynchronously on a shared array buffer and
returns an object { async: bool, value: Promise }. Compare to Atomics.wait(),
waitAsync is non-blocking and usable on the main thread.

TC39 Spec: <a href="https://tc39.es/proposal-atomics-wait-async/">https://tc39.es/proposal-atomics-wait-async/</a>
TC39 Proposal: <a href="https://github.com/tc39/proposal-atomics-wait-async">https://github.com/tc39/proposal-atomics-wait-async</a>
MDN Web Doc: <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/waitAsync">https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/waitAsync</a>

* JSTests/stress/settimeout-starvation.js: Added.
(let.promise.new.Promise):
(wait):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/AtomicsObject.cpp:
(JSC::atomicsWaitImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/Intrinsic.cpp:
(JSC::intrinsicName):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSArrayBufferView.h:
(JSC::JSArrayBufferView::hasArrayBuffer const):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/SimpleTypedArrayController.cpp:
(JSC::SimpleTypedArrayController::isAtomicsWaitAsyncAllowedOnCurrentThread):
* Source/JavaScriptCore/runtime/SimpleTypedArrayController.h:
* Source/JavaScriptCore/runtime/TypedArrayController.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::~VM):
* Source/JavaScriptCore/runtime/WaiterListsManager.h: Added.
(JSC::Waiter::Waiter):
(JSC::Waiter::isAsync const):
(JSC::Waiter::getVM const):
(JSC::Waiter::getPromise const):
(JSC::Waiter::getTicket const):
(JSC::Waiter::dump const):
(JSC::WaiterList::enqueue):
(JSC::WaiterList::dequeue):
(JSC::WaiterList::takeFirst):
(JSC::WaiterList::isEmpty):
(JSC::WaiterList::dump const):
(JSC::WaiterListsManager::singleton):
(JSC::WaiterListsManager::addWaiter):
(JSC::WaiterListsManager::notifyWaiter):
(JSC::WaiterListsManager::timeoutAsyncWaiter):
(JSC::WaiterListsManager::unregister):
(JSC::WaiterListsManager::dump const):
(JSC::WaiterListsManager::RegisteredVMs::add):
(JSC::WaiterListsManager::RegisteredVMs::remove):
(JSC::WaiterListsManager::RegisteredVMs::contains):
(JSC::WaiterListsManager::RegisteredVMs::dump const):
(JSC::WaiterListsManager::notifyWaiterImpl):
(JSC::WaiterListsManager::add):
(JSC::WaiterListsManager::find):
* Source/WebCore/bindings/js/WebCoreTypedArrayController.cpp:
(WebCore::WebCoreTypedArrayController::isAtomicsWaitAsyncAllowedOnCurrentThread):
* Source/WebCore/bindings/js/WebCoreTypedArrayController.h:

Canonical link: <a href="https://commits.webkit.org/257061@main">https://commits.webkit.org/257061@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6429e921f3a554c19b5a6448f3135c87158a3899

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97735 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6977 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30903 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107226 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167494 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7391 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35746 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90120 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103877 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103377 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5547 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84366 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32515 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87421 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89176 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75435 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88633 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/969 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84310 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/959 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22106 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28529 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4849 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5781 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44566 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87141 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2214 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41501 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19551 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -466,6 +474,70 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+        JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+        typedArray->possiblySharedBuffer()->addWaiterRecord(accessIndex, promise);
+        dataLogF("Thread <%p> ArrayBuffer* addr: %p\n",  &Thread::current(), typedArray->possiblySharedBuffer());
+        value = JSValue(promise);
```
`typedArray->possiblySharedBuffer()` in `waitAsync` provides different `ArrayBuffer*` from `notify`.

hyjorc1 commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -522,9 +626,17 @@ JSC_DEFINE_HOST_FUNCTION(atomicsFuncNotify, (JSGlobalObject* globalObject, CallF
+        JSGlobalObject* global = promise->globalObject();
+        promise->resolve(global, jsNontrivialString(global->vm(), "ok"_s));
+    }
+
```
`typedArrayView->possiblySharedBuffer()` in `Notify` provides different `ArrayBuffer*` from `waitAsync`.

hyjorc1 commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -522,9 +626,17 @@ JSC_DEFINE_HOST_FUNCTION(atomicsFuncNotify, (JSGlobalObject* globalObject, CallF
     switch (typedArrayView->type()) {
     case Int32ArrayType: {
         int32_t* ptr = jsCast<JSInt32Array*>(typedArrayView)->typedVector() + accessIndex;
+        dataLogF("Thread <%p> typedVector addr: %p\n",  &Thread::current(), jsCast<JSInt32Array*>(typedArrayView)->typedVector());
```
But `waitAsync` and `notify` have the same `typedVector()` address.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
-            [] () { },
-            MonotonicTime::now() + timeout);
+    if (isAsync) {
+        JSObject* obj = constructEmptyObject(getVM(globalObject), globalObject->objectStructureForObjectConstructor());
```
`getVM` is used in headers. In cpp file, you can just use `globalObject->vm()`.
But in this case, we already have `vm` variable. Let's just use it.

Constellation replied:
And rename obj to object. We would like to have descriptive name as much as possible.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+
+        obj->putDirect(vm, vm.propertyNames->async, async);
+        obj->putDirect(vm, vm.propertyNames->value, value);
+        result = obj;
```
Return `obj` early here. `return object;`.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+        obj->putDirect(vm, vm.propertyNames->async, async);
+        obj->putDirect(vm, vm.propertyNames->value, value);
+        result = obj;
+    } else {
```
Remove `else` things. If we return early, we do not need to put these code in else branch.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
         return { };
     }
 
+    JSValue result;
```
If we return earlier, probably this is not necessary.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+    if (isAsync) {
+        JSObject* obj = constructEmptyObject(getVM(globalObject), globalObject->objectStructureForObjectConstructor());
+
+        JSValue async;
```
Let's have `bool async = false`.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+        JSValue value;
+        didPassValidation = WTF::atomicLoad(ptr) == expectedValue;
+        if (!didPassValidation) {
+            async = jsBoolean(false);
```
`async = false`.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+            async = jsBoolean(false);
+            value = vm.smallStrings.notEqualString();
+        } else if (!timeout) {
+            async = jsBoolean(false);
```
`async = false`.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+            async = jsBoolean(false);
+            value = vm.smallStrings.timedOutString();
+        } else {
+            async = jsBoolean(true);
```
`async = true`.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+            });
+        }
+
+        obj->putDirect(vm, vm.propertyNames->async, async);
```
And do `jsBoolean(async)`.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+            value = JSValue(promise);
+            RunLoop::current().dispatchAfter(timeout, [ptr, promise] {
+                WaiterListsManager::singleton().timeoutAsyncWaiter(ptr, &promise->globalObject()->vm());
+            });
```
Who holds `promise`? (It is GC-managed cell, capturing it in a lambda does not keep it).
Let's not capture `promise` in this lambda.
We should just pass `RefPtr<VM>` to keep VM.

```
            RunLoop::current().dispatchAfter(timeout, [ptr, vm = Ref { vm }] {
                 WaiterListsManager::singleton().timeoutAsyncWaiter(ptr, vm.ptr());
             });
```

hyjorc1 replied:
Before we send the lambda to the run loop, `WaiterListManager` will `addWaiter` for that `promise`. When we create a`Waiter`, we will append a deferred work with a `ticket` to the `vm` to keep it alive until the `ticket` is consumed by `notifyWaiterImpl`.
```
    Waiter(JSPromise* promise)
        : m_isAsync(true)
    {
        VM& vm = promise->globalObject()->vm();
        vm.apiLock().lock();
        data.ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
        vm.apiLock().unlock();
    }
``` 

Where `notifyWaiterImpl` can be only triggered by two cases:

1. The `timeout` function invoked by the callback function passed to the run loop of the `vm` (this is done by `addWaiter` function).
```
        RunLoop::current().dispatchAfter(timeout, [ptr, waiter] { // <-- this waiter is only used as an ID for finding the actual waiter in the queue.
            WaiterListManager::singleton().timeout(ptr, waiter);
        });
```

2. Some other threads call `notifyWaiter` and then `notifyWaiterImpl ` will check whether the `vm` is still alive by
```
        if (!waiter || !m_registeredVMs.contains(waiter->getVM()))
            return false;
```
This is safe even when the `vm` is destructed since we `unregister` VMs in their destructors.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+        }
+
+    private:
+        mutable Mutex lock;
```
Use `Lock` instead of `Mutex`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+
+        void remove(VM* vm)
+        {
+            MutexLocker locker(lock);
```
Use `LockHolder locker { lock }`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+
+        bool contains(VM* vm)
+        {
+            MutexLocker locker(lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+
+        void dump(PrintStream& out) const
+        {
+            MutexLocker locker(lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+    public:
+        void add(VM* vm)
+        {
+            MutexLocker locker(lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+
+    WaiterList* find(void* ptr)
+    {
+        MutexLocker locker(lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+
+    RegisteredVMs m_registeredVMs;
+
+    mutable Mutex lock;
```
Use `Lock`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+
+    WaiterList* add(void* ptr)
+    {
+        MutexLocker locker(lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+        RELEASE_ASSERT(queue);
+
+        {
+            MutexLocker locker(queue->lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+                return waiter->isAsync() && waiter->getVM() == vm;
+            });
+            bool isOKString = false;
+            notifyWaiterImpl(ptr, WTFMove(waiter), isOKString);
```
We should pass locker as an argument `AbstractLockHolder&` to represent that this function can be called only when we have a lock.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+        if (!queue)
+            return notified;
+        {
+            MutexLocker locker(queue->lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+                    break;
+                std::unique_ptr<Waiter> waiter = queue->dequeue();
+                bool isOKString = true;
+                if (notifyWaiterImpl(ptr, WTFMove(waiter), isOKString))
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+        WaiterList* queue = add(ptr);
+        RELEASE_ASSERT(queue);
+        {
+            MutexLocker locker(queue->lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+        WaiterList* queue = add(ptr);
+        RELEASE_ASSERT(queue);
+        {
+            MutexLocker locker(queue->lock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+            manager.construct();
+        });
+        return manager;
+    }
```
This should be defined in cpp file. Otherwise, this `static` variable becomes copied in multiple translation units.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -48,6 +50,7 @@ STATIC_ASSERT_IS_TRIVIALLY_DESTRUCTIBLE(AtomicsObject);
     macro(store, Store, 3)                                              \
     macro(sub, Sub, 3)                                                  \
     macro(wait, Wait, 4)                                                \
+    macro(waitAsync, WaitAsync, 4)                                      \
```
We should expose it only when `Options::useAtomicWaitAsync` is true. And we should add that option flag to `OptionsList.h`. (See `useArrayGroupMethod` for example).

hyjorc1 replied:
Still working on this one.

hyjorc1 replied:
Fixed

Constellation commented on Source/JavaScriptCore/runtime/WaiterListsManager.h:
```diff
@@ -0,0 +1,330 @@
+            return true;
+        }
+        return ParkingLot::unparkCount(ptr, 1);
+    }
```
This looks not OK since non-async case is locked by a global lock.
I think we should keep non-async notification the same to the old one. So we should have appropriate mechanism to avoid a global locking for the non-async notification.
Can you explore the implementation of `ParkingLot` etc. to find the right way to do these things?

hyjorc1 replied:
What do you mean `global lock`? Is it `WaiterListManager.lock`? `notifyWaiterImpl` is only locked by `WaiterList.lock` which is required and specified by the spec (see `Atomics.notify` in https://tc39.es/ecma262/#sec-atomics.notify). This is because `WaiterList` is shared by both `Waiter` and `AsyncWaiter`.

```
9. Perform EnterCriticalSection(WL).
10. Let S be RemoveWaiters(WL, c).
11. Repeat, while S is not empty,
    a. Let W be the first agent in S.
    b. Remove W from the front of S.
    c. Perform NotifyWaiter(WL, W).
    d. Set n to n + 1.
12. Perform LeaveCriticalSection(WL).
```

hyjorc1 replied:
In `bool dequeue(const void* address, BucketMode bucketMode, const DequeueFunctor& dequeueFunctor,
    const FinishFunctor& finishFunctor)`of `ParkingLot.cpp`, it uses `Atomic.load` to get the `Bucket` (`queue`) for the given `address` (`ptr`).
    
Then, it uses `bucket->lock.lock();` and `bucket->lock.unlock();` to dequeue the record in the `Bucket`. Should I use the similar trick for my `HashMap` and `WaiterList`?

shvaikalesh commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -498,6 +563,37 @@ JSC_DEFINE_HOST_FUNCTION(atomicsFuncWait, (JSGlobalObject* globalObject, CallFra
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (!typedArrayView->isShared())
+        throwVMTypeError(globalObject, scope, "Typed array for wait/waitAsync/notify must wrap a SharedArrayBuffer."_s);
```
Should it be `return throwVMTypeError` here? If that's the case, it would be nice to add a test that captures that (like incorrect / throwing `callFrame->argument(1)` doesn't overwrite the `"Typed array for wait/waitAsync/notify must wrap a SharedArrayBuffer."` error).

shvaikalesh reviewed (commented):
:cat:

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -55,6 +57,8 @@ STATIC_ASSERT_IS_TRIVIALLY_DESTRUCTIBLE(AtomicsObject);
 FOR_EACH_ATOMICS_FUNC(DECLARE_FUNC_PROTO)
 #undef DECLARE_FUNC_PROTO
 
+    static JSC_DECLARE_HOST_FUNCTION(atomicsFuncWaitAsync);
```
Do not indent here.

shvaikalesh commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+        vm->deferredWorkTimer->scheduleWorkSoon(ticket, [isOKString](DeferredWorkTimer::Ticket ticket) {
+            JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+            JSGlobalObject* globalObject = promise->globalObject();
+            VM& vm = promise->globalObject()->vm();
```
```suggestion
            VM& vm = globalObject->vm();
```

Constellation replied:
Use `promise->vm()`.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -84,6 +88,9 @@ void AtomicsObject::finishCreation(VM& vm, JSGlobalObject* globalObject)
     FOR_EACH_ATOMICS_FUNC(PUT_DIRECT_NATIVE_FUNC)
 #undef PUT_DIRECT_NATIVE_FUNC
 
+    if (Options::useWaitAsync())
```
We should expose it only when
`Options::useWaitAsync()` and `VM is Default`.

shvaikalesh commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+        return notified;
+    }
+
+    void notifyWaiterImpl(void* data, const bool isOKString)
```
nit: If I was authoring this I would have gone with `enum DeferredWorkResult : bool { OK, Timeout }` or something.

shvaikalesh reviewed (commented):
nits

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,26 +449,50 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
+            isAsyncValue = true;
+            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+            WaiterListManager::singleton().addWaiter(ptr, promise, timeout);
+            value = JSValue(promise);
```
You can just assign `value = promise;`.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -442,26 +449,50 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
     {
         ReleaseHeapAccessScope releaseHeapAccessScope(vm.heap);
-        result = ParkingLot::parkConditionally(
+        bool isAtomicsMode = true;
```
Let's add `enum class Mode { Default, JSAtomics };` under `ParkingLot`, and use that enum instead (`ParkingLot::Mode::Default` v.s. `ParkingLot::Mode::JSAtomics`).

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+        m_registeredVMs.add(&vm);
+        RunLoop::current().dispatchAfter(timeout, [ptr, ticket] {
+            WaiterListManager::singleton().timeout(ptr, ticket);
+        });
```
We should not schedule this when timeout is `infinity` (it becomes infinity when timeout is not specified).

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+        VM& vm = promise->globalObject()->vm();
+        vm.apiLock().lock();
+        auto ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
+        vm.apiLock().unlock();
```
Use `JSLockHolder` instead of explicit `lock()` and `unlock()` calls.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+    void addWaiter(void* ptr, JSPromise* promise, Seconds timeout)
+    {
+        RELEASE_ASSERT(ptr && promise);
+        VM& vm = promise->globalObject()->vm();
```
Use `promise->vm()`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+    {
+        RELEASE_ASSERT(ptr && ticket);
+        JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+        VM* vm = &promise->globalObject()->vm();
```
Use `VM& vm = promise->vm();`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+        RELEASE_ASSERT(ptr && ticket);
+        JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+        VM* vm = &promise->globalObject()->vm();
+        if (!m_registeredVMs.contains(vm))
```
And pass `&vm`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+            return;
+
+        ParkingLot::unparkAsync(ptr, ticket, [this](void* data) {
+            bool isOKString = false;
```
What does it mean? I think we should pass a `enum class`, saying `Notified` or `TimedOut`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+    unsigned notifyWaiter(void* ptr, unsigned count)
+    {
+        RELEASE_ASSERT(ptr);
+        bool isAsyncMode = true;
```
Ditto, use enum.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+        bool isAsyncMode = true;
+        unsigned notified = ParkingLot::unparkCount(
+            ptr, count, [this](void* data) {
+                bool isOKString = true;
```
Ditto. use enum.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+    void notifyWaiterImpl(void* data, const bool isOKString)
+    {
+        auto ticket = static_cast<DeferredWorkTimer::Ticket>(data);
+        VM* vm = &jsCast<JSPromise*>(ticket->target())->globalObject()->vm();
```
Use `VM& vm = jsCast<JSPromise*>(ticket->target())->vm();`

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+    void timeout(void* ptr, DeferredWorkTimer::Ticket ticket)
+    {
+        RELEASE_ASSERT(ptr && ticket);
+        JSPromise* promise = jsCast<JSPromise*>(ticket->target());
```
Let's say that VM is already destroyed, then promise / ticket use here is already UAF.

Constellation commented on WebKit.xcworkspace/contents.xcworkspacedata:
```diff
@@ -1,6 +1,9 @@
    version = "1.0">
+   <FileRef
+      location = "group:Source/JavaScriptCore/runtime/WaiterListManager.h">
+   </FileRef>
```
This looks wrong change.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,139 @@
+    private:
+        mutable Lock lock;
+        HashSet<VM*> set;
+    };
```
I don't think this is good abstraction. We should just carefully manage Lock and HashSet in `WaiterListManager`.

Constellation reviewed (commented):
Initial comments.

Constellation commented on Source/WTF/wtf/ParkingLot.cpp:
```diff
@@ -64,7 +65,56 @@ struct ThreadData : public ThreadSafeRefCounted<ThreadData> {
+        ThreadData* sync;
+        void* async;
+    } u;
+};
```
Use `Variant<>` instead.

hyjorc1 replied:
It's a outdated file.

Constellation commented on Source/JavaScriptCore/runtime/AtomicsObject.cpp:
```diff
@@ -427,8 +433,9 @@ JSC_DEFINE_HOST_FUNCTION(atomicsFuncSub, (JSGlobalObject* globalObject, CallFram
+
 template<typename ValueType, typename JSArrayType>
-JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, unsigned accessIndex, ValueType expectedValue, JSValue timeoutValue)
+JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, unsigned accessIndex, ValueType expectedValue, JSValue timeoutValue, const AtomicsWaitType type)
```
`const AtomicsWaitType` => `AtomicsWaitType`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+    return manager;
+}
+
+JSValue WaiterListManager::wait(JSGlobalObject* globalObject, VM& vm, void* ptr, AtomicsWaitValidation validation, Seconds timeout, const AtomicsWaitType type)
```
`const AtomicsWaitType` => `AtomicsWaitType`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+        const TimeWithDynamicClockType& time = MonotonicTime::now() + timeout;
+
+        {
+            LockHolder locker(list->lock);
```
This is not correct. `addSyncWaiter` is once taking a lock. We should not retake it again.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+    if (validation == AtomicsWaitValidation::Pass) {
+        Waiter waiter { &vm };
+        WaiterList* list = addSyncWaiter(ptr, &waiter);
+        const TimeWithDynamicClockType& time = MonotonicTime::now() + timeout;
```
Let's not take a same `list->lock` twice. It is costly. We should take it only once, and do the following sequence in an atomic manner.

```
Waiter waiter { &vm };
TimeWithDynamicClockType time = MonotonicTime::now() + timeout;
WaiterList* list = findOrCreateList(ptr);
{
    Locker locker { list->lock };
    list->addLast(locker, waiter);
    while (waiter.getVM() && time.nowWithSameClock() < time)
        waiter.getCondition().waitUntil(list->lock, time.approximateWallTime());
    RELEASE_ASSERT(!waiter.getVM() || waiter.getVM() == &vm);
    bool didGetDequeued = !waiter.getVM();
    if (didGetDequeued)
        return vm.smallStrings.okString();
    list->take(locker, &waiter);
    return vm.smallStrings.timedOutString();
}
```

And we should pass `const AbstractLocker&` parameter to the function which needs to be called while taking a lock. This is WebKit's convention.
And not use `LockHolder`. Use `Locker { list->lock }`.

And I don't think we need to make it RELEASE_ASSERT while taking a lock. Let's use ASSERT.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+        return vm.smallStrings.notEqualString();
+}
+
+WaiterList* WaiterListManager::addSyncWaiter(void* ptr, Waiter* waiter)
```
This function is not necessary. Let's not take the same lock twice.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+WaiterList* WaiterListManager::addSyncWaiter(void* ptr, Waiter* waiter)
+{
+    WaiterList* list = findOrCreateList(ptr);
+    waiter->getVM()->addElementPtr(ptr);
```
This should be done in the Waiter's constructor.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+    Waiter* waiter = new Waiter(promise);
+    {
+        LockHolder locker(list->lock);
```
Do not take this lock twice. We should take it only once per one wait.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+void WaiterListManager::timeoutAsyncWaiter(void* ptr, Waiter* target)
+{
+    RELEASE_ASSERT(ptr);
```
Use `ASSERT`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+WaiterList* WaiterListManager::findOrCreateList(void* ptr)
+{
+    LockHolder locker(m_waiterListsLock);
+    return m_waiterLists.add(ptr, makeUnique<WaiterList>()).iterator->value.get();
```
This is not great since it is always allocating `makeUnique<WaiterList>()`.
Use `m_waiterLists.ensure` instead.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+        if (!list)
+            continue;
+        {
+            LockHolder locker(list->lock);
```
Use `Locker { list->lock }`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+void WaiterListManager::unregisterVM(VM* vm)
+{
+    LockHolder locker(m_waiterListsLock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+void WaiterListManager::unregisterSharedArrayBuffer(uint8_t* arrayPtr, size_t size)
+{
+    LockHolder locker(m_waiterListsLock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+WaiterList* WaiterListManager::findOrCreateList(void* ptr)
+{
+    LockHolder locker(m_waiterListsLock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+WaiterList* WaiterListManager::findList(void* ptr)
+{
+    LockHolder locker(m_waiterListsLock);
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/VM.cpp:
```diff
@@ -1445,6 +1449,12 @@ void VM::addLoopHintExecutionCounter(const JSInstruction* instruction)
 
+void VM::addElementPtr(void* elementPtr)
+{
+    Locker locker { m_registeredElementPtrsLock };
```
Why do we need this lock? It is added by VM's thread and touched by VM's thread, right?

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+    if (timeout != Seconds::infinity()) {
+        RunLoop::current().dispatchAfter(timeout, [ptr, waiter] {
+            WaiterListManager::singleton().timeoutAsyncWaiter(ptr, waiter);
+        });
```
This is not correct since `waiter` can be destroyed when VM is destroyed before this is called.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+void WaiterListManager::unregisterVM(VM* vm)
+{
+    LockHolder locker(m_waiterListsLock);
+    for (void* ptr : vm->registeredElementPtrs()) {
```
Why do we need to have registeredElementPtrs? How about just iterating all `m_waiterLists`?

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+unsigned WaiterListManager::notifyWaiter(void* ptr, unsigned count)
+{
+    RELEASE_ASSERT(ptr);
```
ASSERT.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+            LockHolder locker(list->lock);
+            while (waiter.getVM() && time.nowWithSameClock() < time)
+                waiter.getCondition().waitUntil(list->lock, time.approximateWallTime());
+            RELEASE_ASSERT(!waiter.getVM() || waiter.getVM() == &vm);
```
Why RELEASE_ASSERT?

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+
+void WaiterListManager::notifyWaiterImpl(const AbstractLocker&, Waiter* waiter, const ResolveResult resolveResult)
+{
+    RELEASE_ASSERT(waiter);
```
ASSERT.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,185 @@
+        });
+    }
+
+    void addLast(Waiter* waiter)
```
Take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,185 @@
+        m_size++;
+    }
+
+    Waiter* takeFirst()
```
Take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,185 @@
+        return take(result);
+    }
+
+    Waiter* takeLast()
```
Take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,185 @@
+        return take(result);
+    }
+
+    Waiter* take(Waiter* waiter)
```
Take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    template<typename Functor>
+    void deleteIf(const Functor& functor)
```
Take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,241 @@
+    WaiterList* list = findList(ptr);
+    if (list) {
+        LockHolder locker(list->lock);
+        notifyWaiterImpl(locker, list->take(target), ResolveResult::Timeout);
```
Ditto.

Constellation commented on Source/WTF/wtf/RunLoop.h:
```diff
@@ -129,7 +128,7 @@ class WTF_CAPABILITY("is current") RunLoop final : public SerialFunctionDispatch
 #endif
 
-    class TimerBase {
+    class TimerBase : public ThreadSafeRefCounted<TimerBase> {
```
I think we should just inherit this only in DispatchTimer.

Constellation commented on Source/WTF/wtf/RunLoop.h:
```diff
@@ -201,10 +200,6 @@ class WTF_CAPABILITY("is current") RunLoop final : public SerialFunctionDispatch
-    class Holder;
-    static ThreadSpecific<Holder>& runLoopHolder();
-
     class DispatchTimer final : public TimerBase {
```
Change it to `class DispatchTimer final : public TimerBase, public ThreadSafeRefCounted<DispatchTimer>`.

Constellation commented on Source/WTF/wtf/RunLoop.h:
```diff
@@ -129,7 +128,7 @@ class WTF_CAPABILITY("is current") RunLoop final : public SerialFunctionDispatch
 
-    class TimerBase {
+    class TimerBase : public ThreadSafeRefCounted<TimerBase> {
         WTF_MAKE_FAST_ALLOCATED;
```
Remove this and add this to the each final TimerBase classes (not for DispatchTimer since ThreadSafeRefCounted<> has one).

Constellation commented on Source/WTF/wtf/RunLoop.cpp:
```diff
@@ -169,15 +170,16 @@ void RunLoop::dispatch(Function<void()>&& function)
+    timer->setFunction([timer = timer.copyRef(), function = WTFMove(function)] {
         function();
-        delete timer;
+        timer->stop();
```
Let's also explicitly release the timer. Like,
```
Ref<DispatchTimer> protected { WTFMove(timer) };
function();
protected->stop();
```
Then, after the function call, protected will be released.

hyjorc1 replied:
This `timer` is `const`. Can I do `Ref<DispatchTimer> protectedTimer { timer.copyRef() };`?

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,256 @@
+
+                        VM& vm = *waiter->getVM();
+                        {
+                            JSLockHolder locker(vm);
```
This looks like it can cause deadlock.
We should not take a lock of VM. We should do similar thing to `scheduleWorkSoon`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,256 @@
+
+void WaiterListManager::unregisterSharedArrayBuffer(uint8_t* arrayPtr, size_t size)
+{
+    Locker locker { m_waiterListsLock };
```
While taking this lock, I think we can ensure that VMs are alive.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,256 @@
+        });
+
+        if (resolveResult == ResolveResult::Ok)
+            waiter->cancelTimer();
```
is it safe?

hyjorc1 replied:
Yes, I believe so. The `timer` is created and added to `RunLoop` with `CFRunLoopAddTimer` initially. Then `cancelTimer` will stop the `timer` with `CFRunLoopTimerInvalidate`. Since `waiterList` is locked only one thread can access and remove the waiter, and then call `cancelTimer`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,256 @@
+        return object;
+    }
+
+    if (validation == AtomicsWaitValidation::Pass) {
```
You can flip.

```
if (validation == AtomicsWaitValidation::Fail)
    return vm.smallStrings.notEqualString();

Waiter waiter { &vm };
...
```

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,226 @@
+        if (!m_timer)
+            return;
+        m_timer->stop();
+        m_timer->deref();
```
This is a memory corruption bug. `RefPtr` does automatic `ref` / `deref`. If you call it manually, it broke ref count and corrupting memory.
Let's make it `m_timer = nullptr;`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,226 @@
+    {
+        VM& vm = promise->vm();
+        {
+            JSLockHolder locker(vm);
```
This looks incorrect. We should already have a lock of this, so this is not necessary.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,226 @@
+    Waiter* takeFirst(const AbstractLocker&)
+    {
+        return take(&*m_waiters.begin());
+    }
```
This looks wrong. Who is maintaining the ownership of the returned Waiter? It seems that it is releasing memory before returning.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,226 @@
+    Waiter* takeLast(const AbstractLocker&)
+    {
+        return take(&*(--m_waiters.end()));
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,226 @@
+    Waiter* take(const AbstractLocker&, Waiter* waiter)
+    {
+        return take(waiter);
+    }
```
Ditto

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -212,6 +212,24 @@ bool DeferredWorkTimer::cancelPendingWork(Ticket ticket)
+    }
+
+    return result;
+}
```
So, the question is why is it OK to be called from the different thread?

hyjorc1 replied:
The purpose of creating and using `DeferredWorkTimer::cancelAndRemovePendingWork` is to stop the vm hanging on the deferred work which never be scheduled.

Constellation replied:
I think this does not work when it is called from different thread.
Did you check `ticket->cancel` implementation?

Constellation commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -553,6 +553,7 @@ bool canUseWebAssemblyFastMemory();
     v(Bool, useSharedArrayBuffer, false, Normal, nullptr) \
     v(Bool, useShadowRealm, false, Normal, "Expose the ShadowRealm object.") \
     v(Bool, useTemporal, false, Normal, "Expose the Temporal object.") \
+    v(Bool, useWaitAsync, true, Normal, "Expose the waitAsync() methods on Atomics.") \
```
Use name `useAtomicsWaitAsync`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,252 @@
+
+    Waiter waiter { &vm };
+    WaiterList* list = findOrCreateList(ptr);
+    const TimeWithDynamicClockType& time = MonotonicTime::now() + timeout;
```
Do not use const reference here, it is incorrect, right? Just having reference is wrong in C++.
You should use `MonotonicTime time = MonotonicTime::now() + timeout`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,220 @@
+        return m_isAsync;
+    }
+
+    VM* getVM()
```
We do not add `get` name to getters. Use `vm()`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,220 @@
+        m_ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
+    }
+
+    bool isAsync()
```
Put `const`. `bool isAsync() const`

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,220 @@
+        m_vm = nullptr;
+    }
+
+    Condition& getCondition()
```
Ditto.

Constellation replied:
Still using `get` name here.

hyjorc1 replied:
Sorry, forgot that one.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,220 @@
+        return m_condition;
+    }
+
+    DeferredWorkTimer::Ticket getTicket()
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,252 @@
+        if (timeout != Seconds::infinity()) {
+            Ref<RunLoop::DispatchTimer> timer = RunLoop::current().dispatchAfter(timeout, [this, ptr, waiter = waiter.copyRef()] {
+                WaiterList* list = findList(ptr);
+                if (list) {
```
I think this is not allowed since at this point there is no guarantee that we are still keeping SharedArrayBuffer.
What happens if this thread is no longer running JSC VM, and just after running `findList`, some other thread destroyed SharedArrayBuffer?
Given that at this point we have no guarantee that we are still having SharedArrayBuffer alive, we need to be super careful about these conditions.

Constellation replied:
I think WaiterList should be `ThreadSafeRefCounted`. And we should keep ref of that when returning it from `findList` etc. So, `findList` returns `RefPtr<WaiterList>`. And `findOrCreateList` returns `Ref<WaiterList>`

Constellation replied:
And this means that we may find that this waiter is not in the list.

1. Until this, SharedArrayBuffer gets destroyed.
2. New SharedArrayBuffer is created and it has the same address
3. So, when getting a list from this pointer here, we get a new list, which does not include this waiter.

We need to handle this kind of case too.

Constellation commented on Source/WTF/wtf/RunLoop.cpp:
```diff
@@ -169,15 +170,17 @@ void RunLoop::dispatch(Function<void()>&& function)
-    timer->setFunction([timer, function = WTFMove(function)] {
+    Ref<DispatchTimer> timer = adoptRef(*new DispatchTimer(*this));
+    timer->setFunction([timer = timer.copyRef(), function = WTFMove(function)] {
+        Ref<DispatchTimer> protectedTimer { timer.copyRef() };
```
Move timer to this. Not copying. Otherwise, this function is still keeping timer ref even after execution of this lambda.
This WTFMove and protectedTimer is to delete this lambda -> timer reference after running.
```
Ref<DispatchTimer> protectedTimer { WTFMove(timer); };
```
The reason of this is avoiding creating cyclic reference as much as possible to avoid memory leak.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,229 @@
+        return m_condition;
+    }
+
+    DeferredWorkTimer::Ticket getTicket() const
```
Ditto.

hyjorc1 replied:
Sorry, forgot that one.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,229 @@
+
+    VM* vm()
+    {
+        return m_isAsync ? &jsCast<JSPromise*>(m_ticket->target())->vm() : m_vm;
```
Why? We should always set m_vm.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,229 @@
+
+    void nullOutVM()
+    {
+        ASSERT(!m_isAsync);
```
Why?

hyjorc1 replied:
In the initial design, `m_vm` is only used for SyncWaiter and we `nullOutVM` to indicate the SyncWaiter is removed from the list.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,229 @@
+        m_waiters.append(waiter);
+        // Since SyncWaiter will be allocated on stack, we only need to ref AsyncWaiter.
+        if (waiter->isAsync())
+            waiter->ref();
```
I think we can do simpler thing. Let's make sync Waiter allocated from Heap instead of Stack. And placing it under VM's member. And before chaining it into WaiterList, let's initialize it appropriately. Then we do not need to have these special conditions.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,229 @@
+        // `take` function is used to consume a waiter (either notify or timeout).
+        // First, we need make sure it's not removed from the list.
+        ASSERT(waiter->vm() || waiter->getTicket());
+        Ref<Waiter> protectedWaiter = Ref { *waiter };
```
If it is always non nullptr, parameter should be `Waiter&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,229 @@
+        m_waiters.remove(waiter);
+        // Since SyncWaiter will be allocated on stack, we only need to deref AsyncWaiter.
+        if (waiter->isAsync())
+            waiter->deref();
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,229 @@
+        return take(&*m_waiters.begin());
+    }
+
+    Ref<Waiter> take(const AbstractLocker&, Waiter* waiter)
```
If it is always non nullptr, then we should take reference (`Waiter&`).

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+        return;
+    }
+
+    // If waiter is a SyncWaiter, we null out it's vn to indicate that this waiter
```
`vn` -> `vm`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+                    // it shouldn't been blocked or have any deferred work left.
+                    // That means we shouldn't find any SyncWaiter or AsyncWaiter
+                    // associating to the vm. Maybe we should throw error or add assertion here?
+                    if (waiter->isAsync())
```
It should be `ASSERT(waiter->isAsync())`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+                    if (waiter->isAsync())
+                        cancelAsyncWaiter(waiter);
+                    else
+                        UNREACHABLE();
```
Remove them.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+            if (list) {
+                Locker listLocker { list->lock };
+                list->removeIf(listLocker, [&](Waiter* waiter) {
+                    if (waiter->isAsync()) {
```
`ASSERT(waiter->isAsync());`

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+                            return false;
+                        cancelAsyncWaiter(waiter);
+                    } else
+                        UNREACHABLE();
```
Remove this. And not use `UNREACHABLE()`. Use `RELEASE_ASSERT_NOT_REACHED();`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+                });
+            }
+
+            return !list->size();
```
It should be always true.
Thus,
```
ASSERT(!list->size());
return true;
```

Constellation replied:
Discussed offline, we must remove waiterlist always.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+{
+    Locker waiterListsLocker { m_waiterListsLock };
+    WaiterList* list = m_waiterLists.ensure(ptr, [] {
+        return makeUnique<WaiterList>(); 
```
This is not right. Do not allocate RefCounted objects with `makeUnique`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+    Locker waiterListsLocker { m_waiterListsLock };
+    WaiterList* result = nullptr;
+    auto it = m_waiterLists.find(ptr);
+    if (it != m_waiterLists.end())
```
We should early return

```
if (it == m_waiterLists.end())
    return nullptr;
return it->value;
```

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+RefPtr<WaiterList> WaiterListManager::findList(void* ptr)
+{
+    Locker waiterListsLocker { m_waiterListsLock };
+    WaiterList* result = nullptr;
```
This is not necessary.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,229 @@
+    RefPtr<WaiterList> findList(void* ptr);
+
+    Lock m_waiterListsLock;
+    HashMap<void*, std::unique_ptr<WaiterList>> m_waiterLists;
```
`std::unique_ptr` is wrong. It is ref-counted. Then you should use `Ref<WaiterList>`

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -212,6 +212,24 @@ bool DeferredWorkTimer::cancelPendingWork(Ticket ticket)
+bool DeferredWorkTimer::cancelAndRemovePendingWork(Ticket ticket)
+{
+    ASSERT(m_pendingTickets.contains(ticket));
+    ASSERT(ticket->isCancelled() || ticket->vm().currentThreadIsHoldingAPILock() || (Thread::mayBeGCThread() && ticket->vm().heap.worldIsStopped()));
```
This condition is saying that this cannot be called from different thread.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+                        // timeout. Otherwise un-task it.
+                        if (waiter->hasTimer())
+                            return false;
+                        cancelAsyncWaiter(waiter);
```
Discussed offline. We should not allow keeping waiters in the list here. We must remove waiters always.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,289 @@
+                    // cancel the corresponding timer, we should not notify it.
+                    if (waiter->getTicket())
+                        notifyWaiterImpl(listLocker, list->take(listLocker, waiter.ptr()), ResolveResult::Timeout);
+                }
```
This is not correct since waiterlist is removed because SharedArrayBuffer is destroyed, and it can be recreated with the same address range. Thus, it is possible that list can be different one.
```
            Ref<RunLoop::DispatchTimer> timer = RunLoop::current().dispatchAfter(timeout, [this, ptr, waiter = waiter.copyRef()] {
                if (RefPtr<WaiterList> list = findList(ptr)) {
                    Locker listLocker { list->lock };
                    if (waiter->ticket()) {
                        if (waiter->isOnList()) {
                            notifyWaiterImpl(listLocker, list->take(listLocker, waiter), ResolveResult::Timeout);
                            return;
                        }
                    }
                }
                ASSERT(!waiter->isOnList());
                if (waiter->ticket())
                    notifyWaiterImpl(NoLockingNecessary, waiter, ResolveResult::Timeout);
            });
```

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+                        return;
+                    }
+                    notifyWaiterImpl(NoLockingNecessary, WTFMove(waiter), ResolveResult::Timeout);
+                }
```
This is not correct. Who ensures that waiter->ticket() is still valid while we do not have any locks?
Ditto for `isOnList`. Why is it safe to check this condition when it is not taking a lock of the WaiterList?
Please check my comment's code carefully.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,235 @@
+        return m_ticket;
+    }
+
+    DeferredWorkTimer::Ticket nullOutTicket()
```
`takeTicket` would be better name. And should take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+    if (waiter->isAsync()) {
+        VM& vm = *waiter->vm();
+        auto ticket = waiter->ticket();
+        ASSERT(ticket);
```
Why not using `waiter->takeTicket()` to take and null out the ticket field?

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+
+        // If waiter is an AsyncWaiter, we null out its ticket first to indicate that it's notified.
+        // Then, cancel its RunLoop timer if it's not timed-out.
+        waiter->nullOutTicket();
```
And remove this.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+void WaiterListManager::cancelAsyncWaiter(Waiter* waiter)
+{
+    ASSERT(waiter->isAsync());
+    DeferredWorkTimer::Ticket ticket = waiter->nullOutTicket();
```
takeTicket.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+    return size;
+}
+
+void WaiterListManager::cancelAsyncWaiter(Waiter* waiter)
```
Should take `const AbstractLocker&` parameter.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+                // If the vm is about destructing, then it shouldn't
+                // been blocked. That means we shouldn't find any SyncWaiter.
+                ASSERT(waiter->isAsync());
+                cancelAsyncWaiter(waiter);
```
Should pass `listLocker`. This is ok to call it only when we have a lock of the waiterlist.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+                // If the AsyncWaiter has valid timer, then let it
+                // timeout. Otherwise un-task it.
+                if (!waiter->hasTimer())
+                    cancelAsyncWaiter(waiter);
```
Pass `listLocker`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+                ASSERT(waiter->isAsync());
+                // If the AsyncWaiter has valid timer, then let it
+                // timeout. Otherwise un-task it.
+                if (!waiter->hasTimer())
```
We should pass `listLocker` to the parameter of `hasTimer` too.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,235 @@
+        m_timer = WTFMove(timer);
+    }
+
+    bool hasTimer()
```
We should take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+        // Then, cancel its RunLoop timer if it's not timed-out.
+        waiter->nullOutTicket();
+        if (resolveResult != ResolveResult::Timeout)
+            waiter->cancelTimer();
```
We should pass listLocker.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+    return notified;
+}
+
+void WaiterListManager::notifyWaiterImpl(const AbstractLocker&, Ref<Waiter>&& waiter, const ResolveResult resolveResult)
```
Name `const AbstractLocker&` as `listLocker`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,298 @@
+    ASSERT(waiter->isAsync());
+    DeferredWorkTimer::Ticket ticket = waiter->nullOutTicket();
+    waiter->vm()->deferredWorkTimer->scheduleWorkSoon(ticket, [](DeferredWorkTimer::Ticket) mutable { });
+    waiter->cancelTimer();
```
Pass listLocker.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,235 @@
+        return !!m_timer;
+    }
+
+    void cancelTimer()
```
Should take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,235 @@
+        return std::exchange(m_ticket, nullptr);
+    }
+
+    void setTimer(Ref<RunLoop::DispatchTimer>&& timer)
```
Should take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,235 @@
+        return m_condition;
+    }
+
+    DeferredWorkTimer::Ticket ticket() const
```
Should take `const AbstractLocker&`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,235 @@
+        m_vm = vm;
+    }
+
+    void nullOutVM()
```
Should take `const AbstractLocker&`. And rename it to `clearVM`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -0,0 +1,235 @@
+            }
+        }
+        return false;
+    }
```
This looks wrong design. The best part of linked-list is `O(1)` removal. So we should not do `O(N)` removal here.
```
bool findAndRemove(const AbstractLocker&, Waiter& target)
{
    if (!target.isOnList())
        return false;
    removeWithUpdate(target);
    return true;
}
```

Constellation replied:
You can add test to ensure that this entry is on this list, but we should not do it in release build.
```
#if ASSERT_ENABLED
if (target.isOnList()) {
    bool found = false;
    for (auto iter = m_waiters.begin(); iter != m_waiters.end(); ++iter) {
        if (&*iter == &target)
            found = true;
    }
    ASSERT(found);
}
#endif
```

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,293 @@
+JSValue WaiterListManager::wait(JSGlobalObject* globalObject, VM& vm, void* ptr, AtomicsWaitValidation validation, Seconds timeout, AtomicsWaitType type)
+{
+    if (type == AtomicsWaitType::Async) {
+        JSObject* object = constructEmptyObject(globalObject->vm(), globalObject->objectStructureForObjectConstructor());
```
`objectStructureForObjectConstructor` is not necessary.
Use `constructEmptyObject(globalObject)`.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,293 @@
+
+        if (timeout != Seconds::infinity()) {
+            Ref<RunLoop::DispatchTimer> timer = RunLoop::current().dispatchAfter(timeout, [this, ptr, waiter = waiter.copyRef()]() mutable {
+
```
Remove this empty line.

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -0,0 +1,293 @@
+    Ref<WaiterList> list = m_waiterLists.ensure(ptr, [] {
+        return adoptRef(*new WaiterList()); 
+    }).iterator->value.get();
+    return list;
```
We do not need to take this result as `list` variable.
```
return m_waiterLists.ensure(ptr, [] {
    return adoptRef(*new WaiterList()); 
}).iterator->value.get();
```

Constellation reviewed (approved):
r=me with comments

hyjorc1 committed:
[JSC] Add support for Atomics.waitAsync
https://bugs.webkit.org/show_bug.cgi?id=241414
rdar://94655073

Reviewed by Yusuke Suzuki.

Atomics.waitAsync() waits asynchronously on a shared array buffer and
returns an object { async: bool, value: Promise }. Compare to Atomics.wait(),
waitAsync is non-blocking and usable on the main thread.

TC39 Spec: https://tc39.es/proposal-atomics-wait-async/
TC39 Proposal: https://github.com/tc39/proposal-atomics-wait-async
MDN Web Doc: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/waitAsync

* JSTests/stress/settimeout-starvation.js: Added.
(let.promise.new.Promise):
(wait):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/AtomicsObject.cpp:
(JSC::atomicsWaitImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/Intrinsic.cpp:
(JSC::intrinsicName):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSArrayBufferView.h:
(JSC::JSArrayBufferView::hasArrayBuffer const):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/SimpleTypedArrayController.cpp:
(JSC::SimpleTypedArrayController::isAtomicsWaitAsyncAllowedOnCurrentThread):
* Source/JavaScriptCore/runtime/SimpleTypedArrayController.h:
* Source/JavaScriptCore/runtime/TypedArrayController.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::~VM):
* Source/JavaScriptCore/runtime/WaiterListsManager.h: Added.
(JSC::Waiter::Waiter):
(JSC::Waiter::isAsync const):
(JSC::Waiter::getVM const):
(JSC::Waiter::getPromise const):
(JSC::Waiter::getTicket const):
(JSC::Waiter::dump const):
(JSC::WaiterList::enqueue):
(JSC::WaiterList::dequeue):
(JSC::WaiterList::takeFirst):
(JSC::WaiterList::isEmpty):
(JSC::WaiterList::dump const):
(JSC::WaiterListsManager::singleton):
(JSC::WaiterListsManager::addWaiter):
(JSC::WaiterListsManager::notifyWaiter):
(JSC::WaiterListsManager::timeoutAsyncWaiter):
(JSC::WaiterListsManager::unregister):
(JSC::WaiterListsManager::dump const):
(JSC::WaiterListsManager::RegisteredVMs::add):
(JSC::WaiterListsManager::RegisteredVMs::remove):
(JSC::WaiterListsManager::RegisteredVMs::contains):
(JSC::WaiterListsManager::RegisteredVMs::dump const):
(JSC::WaiterListsManager::notifyWaiterImpl):
(JSC::WaiterListsManager::add):
(JSC::WaiterListsManager::find):
* Source/WebCore/bindings/js/WebCoreTypedArrayController.cpp:
(WebCore::WebCoreTypedArrayController::isAtomicsWaitAsyncAllowedOnCurrentThread):
* Source/WebCore/bindings/js/WebCoreTypedArrayController.h:

Canonical link: https://commits.webkit.org/257061@main
